### PR TITLE
Fix bugs with streaming Reader

### PR DESCRIPTION
**Pull request by dsnet**
* If the return code is BROTLI_RESULT_NEEDS_MORE_OUTPUT, we do not need to fill
the input buffer again. Otherwise, we risk premature termination since we hit
io.EOF in the underlying Reader first.
* We must check that the input buffer is not empty. Otherwise dereference of
&p[0] will panic.
* We must convert an io.EOF received from the underlying Reader into an
io.ErrUnexpectedEOF. The only valid EOF condition is when the decompressor tells
us so with a BROTLI_RESULT_SUCCESS code.
* Brotli documentation does not guarantee that calling BrotliStateCleanup twice
is okay. Thus, the use of a Finalizer or the fact that someone may call Close
multiple times is not okay.
* Reader errors should persist across multiple Read calls.
* Close should not close the underlying Reader. I don't know of any standard
library Reader that does this. Most certainly the compression library do not
do this.
* After Close, it should be invalid to call Read.

### CODE REVIEW

**Comment by kothar** [comment]
Gah, didn't realise you were still working on this, so I've made conflicting changes. Looking into merging now.

**Comment by dsnet** [comment]
I'm pretty much done, feel free to manually merge the changes you want and I'll close this PR down.

File: dec/decode.go [hunk 1]
```diff
@@ -169,10 +178,11 @@ func NewBrotliReader(stream io.Reader) *BrotliReader {
 func NewBrotliReaderSize(stream io.Reader, size int) *BrotliReader {
 	r := &BrotliReader{
 		reader: stream,
+		state:  new(C.BrotliState),
```
Comment (kothar) [inline dec/decode.go]: This is presumably still going to cause GC panics (#14) when the C struct is traversed?
Should we replace this with []byte and unsafe pointers?

Comment (dsnet) [inline dec/decode.go]: I would just wait and see what people say on the forums.

The example from the youtube repo uses the []byte method, but the [Go programming book written by one of the Go developers](https://github.com/adonovan/gopl.io/blob/master/ch13/bzip/bzip2.go) uses the method you have now.

**Comment by kothar** [comment]
OK, I think that's all working - round trip tests with streaming compression/decompression seem to be successful.

**Pull request merged**